kwasm-lemmas.md

```diff
@@ -308,68 +310,51 @@ We want to make the variant explicit, so we introduce the following helper, whic
 ```
 
-From the semantics, it should be clear that setting the index in a bytemap to the value already contained there will leave the map unchanged.
-Conversely, setting an index in a map to a value `VAL` and then retrieving the value at that index will yield `VAL`.
-A getting operation can safeley ignore setting operation on unrelated indices.
-
-```k
-    rule #set(BMAP, IDX, #get(BMAP, IDX)) => BMAP [smt-lemma]
-
-    rule #get(#set(BMAP, IDX', VAL), IDX) => VAL             requires         IDX ==Int IDX'
-    rule #get(#set(BMAP, IDX', VAL), IDX) => #get(BMAP, IDX) requires notBool IDX ==Int IDX'
-```
-
 TODO: We should inspect the two functions `#getRange` and `#setRange` closer.
 They are non-trivial in their implementation, but the following should obviously hold from the intended semantics.
 
 ```k
-    rule #setRange(BM, EA, #getRange(BM, EA, WIDTH), WIDTH) => BM
-    rule #setRange(BM, EA, VAL, WIDTH) => #set(BM, EA, #wrap(1, VAL))
-      requires WIDTH ==Int 1
-      [simplification]
-
-    rule #getRange(BM, ADDR, WIDTH) modInt BYTE_SIZE => #get(BM, ADDR)
-      requires BYTE_SIZE ==Int 256
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) >>Int SHIFT => #getRange(BM, ADDR +Int 1, WIDTH -Int 1) >>Int (SHIFT -Int 8)
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool 8 <=Int SHIFT
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #get(BM, ADDR)
-      requires N ==Int 1
-       andBool notBool (WIDTH <=Int 0)
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH)
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool 2 ^Int (8 *Int WIDTH) <=Int MAX
        andBool #isByteMap(BM)
       [simplification]
 
-    rule #wrap(N, #getRange(BM, ADDR, WIDTH)) => #getRange(BM, ADDR, N)
-      requires 0 <=Int N andBool N <=Int WIDTH
+    rule #getRange(BM, ADDR, WIDTH) modInt MAX => #getRange(BM, ADDR, WIDTH -Int 1) modInt MAX
+      requires 0 <Int MAX andBool 0 <Int WIDTH
+       andBool MAX <=Int 2 ^Int (8 *Int (WIDTH -Int 1))
+       andBool #isByteMap(BM)
       [simplification]
 
-    rule #getRange(BM, ADDR, WIDTH) => #get(BM, ADDR)
-      requires WIDTH ==Int 1
+    rule #wrap(MAX_WIDTH, #getRange(BM, ADDR, WIDTH)) => #getRange(BM, ADDR, minInt(MAX_WIDTH, WIDTH))
+      requires #isByteMap(BM)
       [simplification]
 ```
 
 `#getRange` over `#setRange`
 
 ```k
+    rule #setRange(BM, ADDR, #getRange(BM, ADDR, WIDTH), WIDTH) => BM [simplification]
+    rule #setRange(BM, ADDR, VAL1 +Int (VAL2 <<Int SHIFT), WIDTH)
+      => #setRange(#setRange(BM, ADDR, VAL1, minInt(#byteWidth(VAL1), WIDTH)), ADDR +Int #byteWidth(VAL1), VAL2, WIDTH -Int #byteWidth(VAL1))
+      requires 0 <=Int ADDR
+       andBool 0  <Int WIDTH
+       andBool #byteWidth(VAL1) *Int 8 ==Int SHIFT
+      [simplification]
+
```

Comment:
Does the proof go thorugh even without the `[simplification]` attribute? If so, best to not add it, since the RHS is actually larger.

Reply:
I think I still want this to apply during simplification because it enables `#getRange(#setRange(#setRange(...)))` lemmas to apply.

If we just consider the `<<Int` to be higher complexity than `minInt`, then the RHS is simpler than the LHS.

Often times tools which try to prove termination will assign a complexity polynomial to each symbol that shows up, and show that the polynomial associated with the RHS of a rule is lower than that associated with the LHS of the rule. In this case, we just need to assign a higher exponent to `<<Int` than to `minInt` to show that we are terminating with this rule.

Reply:
Fair enough. I've looked at working out a general lexicographic order for the lemmas, but that's outdated now. But assigning a high exponent to the shifting rules was part of it, so this seems fine.